PROPOSAL.md

```diff
@@ -0,0 +1,336 @@
+
+1. Tell the Server what we have: nothing.
+2. The server enumerates its current _index_: the UUIDs and versions (hashes) of
+   all source metadata, plus the current global version in the [`ETag`
+   header][etag].
+   - Each source's metadata includes a `collection_version` property over its
+     `collection` of submissions and replies, also enumerated by UUID and
+     version.
+3. Ask the server for what we're missing: everything.
+4. The server returns the metadata for all sources and their _collections_
+   (submissions and replies).
+
+[etag]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/ETag
+
+```mermaid
+sequenceDiagram
+autonumber
+
+participant Client
+participant Server
+
+Client ->> Server: GET /head
+Server ->> Client: 82861 bytes<br>ETag: <version>
+
+Client ->> Server: POST /index (29865 bytes)
+Server ->> Client: 4855285 bytes
+```
+
+|         | Data    | Time      |
+| ------- | ------- | --------- |
+| Total   | 4.97 MB | 70.85 sec |
+| Speedup | 0.98    | 0.69      |
+
+#### Something's changed
+
+If _any_ server-side state has changed, whether from a source or journalist (or
+via `loaddata.py`):
+
+1. Tell the Server what we have: some _version_, which is the SHA-256 hash of
+   our own _index_.
+2. The server enumerates its current _index_: it has a different _version_, so
+   it sends us the entire _index_.
+3. Ask the server for what we're missing: here, it's 3 new sources.
+4. The server returns the metadata for just those sources and their
+   _collections_.
+
+```mermaid
+sequenceDiagram
+autonumber
+
+participant Client
+participant Server
+
+Client ->> Server: GET /head<br>If-None-Match: <version>
+Server ->> Client: 83046 bytes<br>ETag: <version>
+
+Client ->> Server: POST /index (125 bytes)
+Server ->> Client: 11562 bytes
+```
+
+##### Cache miss: server has to recalculate index
+
+|         | Data    | Time      |
+| ------- | ------- | --------- |
+| Total   | 0.08 MB | 29.23 sec |
+| Speedup | 61.00   | 1.68      |
+
+##### Cache hit: server returns cached version
+
+As prototyped here, the cache is invalidated:
+
+- after inserting, updating, or deleting any `Source`, `Submission`, or `Reply`;
+  or
+- after 1 hour.
+
+|         | Data    | Time          |
+| ------- | ------- | ------------- |
+| Total   | 0.08 MB | **18.38** sec |
+| Speedup | 61.00   | **2.68**      |
+
+#### Steady state
+
+Next time, and in fact _most_ times, we get lucky:
+
+1. Tell the Server what we have: again, the _version_ of our current _index_.
+2. The Server's current _index_ has the same _version_, so it returns HTTP
+   `304`: not modified.
+
+```mermaid
+sequenceDiagram
+autonumber
+
+participant Client
+participant Server
+
+Client ->> Server: GET /head<br>If-None-Match: <version>
+Server ->> Client: 20 bytes
+```
+
+|         | Data     | Time     |
+| ------- | -------- | -------- |
+| Total   | 20 bytes | 0.66 sec |
+| Speedup | 244,000  | 74.76    |
+
+## Implementation suggestions
```

Comment:
not sure if we want to call out in the implementation suggestions that the client could eventually batch its requests for source updates? something we talked about briefly